inspirehep/modules/records/validators/helpers.py

```diff
@@ -0,0 +1,88 @@
+#
+# This file is part of Invenio.
+# Copyright (C) 2017 CERN.
+#
+# Invenio is free software; you can redistribute it
+# and/or modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation; either version 2 of the
+# License, or (at your option) any later version.
+#
+# Invenio is distributed in the hope that it will be
+# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Invenio; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Helpers for validators module."""
+
+from __future__ import absolute_import, division, print_function
+
+from invenio_db import db
+from jsonschema import ValidationError
+from collections import deque
+
+FIELD_VALUE_REQUIRE_FIELD_TEMPLATE = "'{field}' field value '{value}' requires field '{required}' to exist."
+FIELD_REQUIRE_FIELD_VALUES_TEMPLATE = "'{field}' field requires '{required}' field to have " \
+                                      "at least one of the values {values}."
+FIELD_REQUIRE_FIELD_TEMPLATE = "'{field}' field requires '{required}' field to exist."
+FIELD_DUPLICATE_VALUES_FOUND_TEMPLATE = "'{field}' field with value '{value}' found in more than one record."
+GET_RESULTS_FOR_FIELD_PROPERTY_QUERYSTRING_TEMPLATE = "SELECT id FROM records_metadata as r, " \
+                                                      "json_array_elements(r.json -> '{field}') " \
+                                                      "as elem WHERE elem ->> '{prop}'='{value}'"
```

Comment:
You're right that we currently can't, but you're not right on the reason why. The problem is that the `json` column in `records_metadata` is implemented as a `JSONType` from `sqlalchemy_utils`. This type doesn't support `__getitem__`, so you can't do queries that look inside of the record for certain values.

In fact, SQLAlchemy 1.0 **has** a JSON type that supports `__getitem__`: https://docs.sqlalchemy.org/en/rel_1_0/dialects/postgresql.html?highlight=json#sqlalchemy.dialects.postgresql.JSON. As you see, this guy comes from the PostgreSQL dialect, which is why it was not used upstream in `invenio_records`: they are also supporting MySQL, but at the time there was no DB-agnostic implementation, so they used the one provided by `sqlalchemy_utils`.